### PR TITLE
[SPARK-8977][Streaming] Defines the RateEstimator interface, and impements the RateController

### PR DESCRIPTION
**Pull request by dragos**
Based on #7471.
- [x] add a test that exercises the publish path from driver to receiver
- [ ] remove Serializable from `RateController` and `RateEstimator`

### CODE REVIEW

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * :: DeveloperApi ::
+ * A StreamingListener that receives batch completion updates, and maintains
+ * an estimate of the speed at which this stream should ingest messages,
+ * given an estimate computation from a `RateEstimator`
+ */
+@DeveloperApi
+private [streaming] abstract class RateController(val streamUID: Int, rateEstimator: RateEstimator)
+  extends StreamingListener with Serializable {
```
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: Probably not worth to serialize (same for RateEstimator). If we make them `@transient`, we should add a test that after recovery from a checkpoint the rate controller is still active (non-null and registered with the streaming listener bus).

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala [hunk 2]
```diff
@@ -47,6 +49,27 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
   /** This is an unique identifier for the input stream. */
   val id = ssc.getNewInputStreamId()
 
+  /**
+   * A rate estimator configured by the user to compute a dynamic ingestion bound for this stream.
+   * @see `RateEstimator`
+   */
+  protected [streaming] val rateEstimator = newEstimator()
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: Why do you need `rateEstimator` as a field in `DStream`? All you need expose to modules outside DStream is the RateController. And the RateEstimator can be within the `RateController`

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala [hunk 3]
```diff
@@ -47,6 +49,27 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
...
+
+  /**
+   * Return the configured estimator, or `noop` if none was specified.
+   */
+  private def newEstimator() =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: Again, this can be all within the RateController, does not need to be a top level field in DStream. In fact it can be something on a `RateEstimator` objects.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala [hunk 4]
```diff
@@ -47,6 +49,27 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
...
+    ssc.conf.get("spark.streaming.RateEstimator", "noop") match {
+      case "noop" => new NoopRateEstimator()
+      case estimator => throw new IllegalArgumentException(s"Unknown rate estimator: $estimator")
+    }
+
+
+  // Keep track of the freshest rate for this stream using the rateEstimator
+  protected[streaming] val rateController: RateController = new RateController(id, rateEstimator) {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: I think the rateController should be a option. By default, it will None. Only when backpressure is enabled do we populate it.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala [hunk 5]
```diff
@@ -34,12 +34,32 @@ import org.apache.spark.{Logging, SparkConf}
...
-  private lazy val rateLimiter = GuavaRateLimiter.create(desiredRate)
+  // treated as an upper limit
+  private val maxRateLimit = conf.getLong("spark.streaming.receiver.maxRate", Long.MaxValue)
+  private lazy val rateLimiter = GuavaRateLimiter.create(maxRateLimit.toDouble)
 
   def waitToPush() {
-    if (desiredRate > 0) {
-      rateLimiter.acquire()
-    }
+    rateLimiter.acquire()
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: Why was this changed?

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala]: Aah, these should not show up in the diff, because the master already has these. Could you update your branch by merging the master?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala (hunk 1 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: Yes. Agreed. Transient plus update some existing test case in CheckpointSuite

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      time: Long,
+      elements: Long,
+      processingDelay: Long,
+      schedulingDelay: Long): Option[Double]
+}
+
+/**
+ * The trivial rate estimator never sends an update
+ */
+private[streaming] class NoopRateEstimator extends RateEstimator {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: There should be no need for such an estimator if Options are used.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.spark.streaming.scheduler.rate
+
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * :: DeveloperApi ::
+ * A component that estimates the rate at wich an InputDStream should ingest
+ * elements, based on updates at every batch completion.
+ */
+@DeveloperApi
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: Does not make sense to have non-public dev API.

**Comment by dragos** [comment]
These tests took a while to debug, so I'll remove the serializable components tonight or tomorrow (it's a pretty localized and isolated change). In the meantime please have a look at the current state.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala [hunk 8]
```diff
@@ -47,6 +49,14 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
   /** This is an unique identifier for the input stream. */
   val id = ssc.getNewInputStreamId()
 
+  // Keep track of the freshest rate for this stream using the rateEstimator
+  protected[streaming] val rateController: Option[RateController] =
+    RateEstimator.makeEstimator(ssc.conf).map { estimator =>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: This should change a little. See comment on what the config params should be in another comment below (near `RateEstimator.makeEstimator`)

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala]: And InputDStream should not have a RateController by default. So this should be None. Its will be upto implementation of InputDStream to override and define.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/ReceiverInputDStream.scala [hunk 9]
```diff
@@ -41,6 +42,12 @@ abstract class ReceiverInputDStream[T: ClassTag](@transient ssc_ : StreamingCont
   extends InputDStream[T](ssc_) {
 
   /**
+   * Asynchronously maintains & sends new rate limits to the receiver through the receiver tracker.
+   */
+  override protected[streaming] val rateController: Option[RateController] =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/ReceiverInputDStream.scala]: This should change a little. See comment on what the config params should be in another comment below (near `RateEstimator.makeEstimator`)

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 10]
```diff
@@ -66,6 +66,12 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
     }
     eventLoop.start()
 
+    // Estimators receive updates from batch completion
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: This comment is not relevant. Shouldnt it be "Attach the rate controllers of the input dstreams for receiving batch info updates"?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    ThreadUtils.newDaemonSingleThreadExecutor("stream-rate-update"))
+
+  private val rateLimit: AtomicLong = new AtomicLong(-1L)
+
+  /**
+   * Compute the new rate limit and publish it asynchronously.
+   */
+  private def computeAndPublish(time: Long, elems: Long, workDelay: Long, waitDelay: Long): Unit =
+    Future[Unit] {
+      val newSpeed = rateEstimator.compute(time, elems, workDelay, waitDelay)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: newSpeed  --> newRate

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      newSpeed foreach { s =>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: We dont generally use postfix notation. use `newSpeed.foraech`

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * Return a new RateEstimator based on the value of `spark.streaming.RateEstimator`.
+   *
+   * @return None if there is no configured estimator, otherwise an instance of RateEstimator
+   * @throws IllegalArgumentException if there is a configured RateEstimator that doesn't match any
+   *         known estimators.
+   */
+  def makeEstimator(conf: SparkConf): Option[RateEstimator] =
+    conf.getOption("spark.streaming.RateEstimator") map { estimator =>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: Let's make the configuration params as follows
`spark.streaming.backpressure.enable = true/false` to enable/disable the whole feature. false by default in 1.5. Also to the docs/configuration.md.
`spark.streaming.backpressure.rateEstimator = pid` to specify which algorithm to use for estimating. DO NOT add this to the docs/configuration.

Basically, the scope `spark.streaming.backpressure` will contain on backpressure related configuration. 

Accordingly, the RateController in InputDStreams will be created as `RateController.create()`, and the estimator will be created by `RateEstimator.create`.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+import org.scalatest.Matchers._
+import org.scalatest.concurrent.Eventually._
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.streaming._
+import org.apache.spark.streaming.scheduler.rate.RateEstimator
+
+
+
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: nit: extra line

**Comment by tdas** [comment]
Time to remove the WIP from the title :) Its looking good. Will take another pass once these comments are addressed.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala (hunk 13 above)
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: The only problem I see is that we'll have different `RateControllers` in different input streams. For instance
-  `ReceiverInputDStream` publishes using the receiver tracker, but the KafkaDirect receiver won't.
- the `ReceiverRateController` needs access to the receiver tracker, so currently it's an inner class of `ReceiverInputDStream`.

This is the main reason I wanted to have `publish` be a method (`protected[streaming]`) on input streams. Usually the logic of *what to do* with the rate is closer to the specific type of stream, than the controller. If we moved `publish` to the stream, we could have a single implementation of `RateController`, that can be controlled by a single boolean `spark.streaming.backpressure.enabled`.

@tdas What do you think?

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: "InputDStream.publish" is not something where its immediately clear from the name on what it means in the context of the InputDstream. In fact, publishing a rate is internal implementation detail of how rate controlling is implemented, and not something that should be fundamentally added as a field to the abstract parent class  InputDStreams. In contrast, the field rateController is something that is more generic and implementation-agnostic and is okay to have as a field in the abstract parent class. 

So I propose this. 
1. Let there be a object method `RateController.isBackpressureEnabled`, which will return true if `spark.streaming.backpressure.enabled`. 
2. Class `ReceiverInputDStream` will create a controller instance of `ReceiverRateController` if the above method returns true. You may or may not wrap that check in an object method, `ReceiverRateController.create(): Option[RateController]`
3. `DirectKafkaRateController` will create a conroller instance of `DirectKafkaRateController` in the same way.

**Comment by tdas** [comment]
Jenkins, test this again.

**Comment by tdas** [comment]
Can you also added an unit test that is verifies that the rate controller is disable by default. That is `ReceiverInputDStream.rateController` is None by default.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    val output = new TestOutputStreamWithPartitions(dstream)
+    output.register()
+    runStreams(ssc, 1, 1)
+
+    eventually(timeout(2.seconds)) {
+      assert(dstream.publishCalls === 1)
+    }
+  }
+
+  test("receiver rate controller updates reach receivers") {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: reach -> each

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+import org.apache.spark.streaming.scheduler.rate.RateEstimator
+
+class RateControllerSuite extends TestSuiteBase {
+
+  override def actuallyWait: Boolean = true
+
+  test("rate controller publishes updates") {
+    val ssc = new StreamingContext(conf, batchDuration)
+    val dstream = new MockRateLimitDStream(ssc, Seq(Seq(1)), 1)
+    val output = new TestOutputStreamWithPartitions(dstream)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: This old style of using `TestSuiteBase` and defining `streamingContext` in every test and making runStreams stop the context is actually very flaky, as failure before `runStream` will end up leaking an active SparkContext and failing all subsequent Spark tests. Its being used in many places and I hope to upgrade them at some point. So the two alternatives of using TestSuiteBase are

1. Use [`usingStreamingContext`](https://github.com/apache/spark/blob/1f6b0b1234cc03aa2e07aea7fec2de7563885238/streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala#L270) - This ensure the context are stopped no matter what fails. But it ends up creating a new SpakrContext for every test. 

2. The 2nd alternative is to share SparkContext and use before-after methods as I have done it in this [suite](https://github.com/apache/spark/blob/master/extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala). 

Please update the testsuites here with either of the two methods.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 17]
```diff
@@ -110,8 +110,15 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
 }
 
-/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
-private class RateLimitInputDStream(@transient ssc_ : StreamingContext)
+/**
+ * An input DStream with a hard-coded receiver that gives access to internals for testing.
+ *
+ * @note Make sure to call {{{SingletonDummyReceiver.reset()}}} before using this in a test,
+ *       or otherwise you may get {{{NotSerializableException}}} when trying to serialize
+ *       the receiver.
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Why does it give that error?

**Comment by tdas** [comment]
Just so that the comment to a hidden thread is not lost, I am repeating it here. For the `rateController: Option[RateController]` I propose this. 
1. Let there be a object method RateController.isBackpressureEnabled, which will return true if spark.streaming.backpressure.enabled. 
2. Class ReceiverInputDStream will create a controller instance of ReceiverRateController if the above method returns true. You may or may not wrap that check in an object method, ReceiverRateController.create(): Option[RateController]
3. DirectKafkaRateController will create a conroller instance of DirectKafkaRateController in the same way.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala (hunk 16 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: Good point about leaking contexts, I'll see what I can do. Since `runStreams` stops the context, we'll stop it twice (I know, it only logs a warning, but still it seems suboptimal).

I'd like to keep using `runStreams`, otherwise I'll have to duplicate some of its implementation. The output testing is collateral, but not bad in itself, at least not worse than duplicating the logic around manual clock advancing. I *do* need to simulate running for a few batch intervals to see how the rate gets updated. 

Lastly, this *particular* test doesn't need it, but I updated to be similar to the others here, and I think it "ties the suit together" :). If you insist I can revert it, but the other two tests, that actually run the whole pipeline to see rate updates in action, still need `runStreams`

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 17 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Because after being `started` once, it has a reference to the ReceiverSupervisor, which is not serializable. So the second time the receiver is used, it can't be serialized and shipped to executors. I thought the one-line method is self-explanatory, but I can add a few more lines of comments around that one line :)

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala (hunk 15 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: It's really "reach". There's only one receiver in my test, so it would be a stretch to describe it that way.

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: right. Its technically correct, but hard to read which is why my brain messed it up. How about "publish rates reach receivers"?

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 17 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Actually, they are already there, in the scaladoc of `SingletonDummyReceiver`, which is linked from this doc.

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Ohh, right. Nvm then.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/ReceiverInputDStream.scala [hunk 18]
```diff
@@ -41,6 +42,16 @@ abstract class ReceiverInputDStream[T: ClassTag](@transient ssc_ : StreamingCont
...
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/ReceiverInputDStream.scala]: nit: Please use `{` and `}` for multiline methods.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+object RateEstimator {
+
+  /**
+   * Return a new RateEstimator based on the value of `spark.streaming.RateEstimator`.
+   *
+   * @return None if there is no configured estimator, otherwise an instance of RateEstimator
+   * @throws IllegalArgumentException if there is a configured RateEstimator that doesn't match any
+   *         known estimators.
+   */
+  def create(conf: SparkConf): Option[RateEstimator] =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: Why should this return an option? Since there is a separate conf that enables/disables RateController, this should always return an instance of RateEstimator even if its the default. But I do understand that since this PR does not have a default one this is probably better. So I am okay to merge this PR as is, as long as the PID estimator PR updates this code to not use Option. Does that make sense?

File: streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala [hunk 20]
```diff
@@ -391,6 +393,32 @@ class CheckpointSuite extends TestSuiteBase {
...
+    SingletonDummyReceiver.reset()
+
+    val output = new TestOutputStreamWithPartitions(dstream.checkpoint(batchDuration * 2))
+    output.register()
+    runStreams(ssc, 5, 5)
+
+    SingletonDummyReceiver.reset()
+    ssc = new StreamingContext(checkpointDir)
+    ssc.start()
+    val outputNew = advanceTimeWithRealDelay(ssc, 2)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala]: Why real delay? Why not just use eventually?

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala (hunk 16 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: Tests that update the rate and checks whether receiver gets the update should not require running the batches. 

If you need runStreams just to "wait" the batches run, then it is cleaner to just use normal clock (no manual clock) and then just use eventually to wait for the necessary condition to be fulfilled. Now, in some places, you are calling runstreams, as well as using eventually.  This change would simplify it further 

```
ssc.start()

eventualy(...) {
}
```
Regarding stopping it twice, if you are anyways adding code in a `try..catch` or `after` to stop the context, might as well use just `usingStreamingContext` or just `after/afterAll` forget about the `TestSuiteBased` and `runStreams`

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+  test("receiver rate controller updates reach receivers") {
+    val ssc = new StreamingContext(conf, batchDuration)
+    withStreamingContext(ssc) { ssc =>
+      val dstream = new RateLimitInputDStream(ssc) {
+        override val rateController =
+          Some(new ReceiverRateController(id, new ConstantEstimator(200.0)))
+      }
+      SingletonDummyReceiver.reset()
+
+      val output = new TestOutputStreamWithPartitions(dstream)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: As I said in the other thread, this `output` stuff is totally superfluous.

Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: It's not, as long as I use `runStreams`. Nothing happens if I don't have any output stream. This test is an end-to-end test, so I need to get batch updates, calculate the rate, and set it back. If there's no output stream, no job gets created.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala (hunk 19 above)
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala]: ok

File: streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala (hunk 20 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala]: I don't remember, the thing is that it didn't get any batch updates without a real delay. I didn't investigate further.

**Comment by dragos** [comment]
@tdas I have some comments in my email that don't appear here, so I suppose you changed your mind and deleted it.

I think the review would be smoother if you explained a bit better what are your underlying concerns, rather than very specific points. For instance, I don't understand why you insist I should avoid using your own helper methods. I don't think code would get clearer, nor less flaky. If anything, reusing these methods makes the code more uniform and easier to understand, plus any bugs that get fixed in those helpers benefit this test suite too.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala (hunk 21 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: You just need to call `register` on the `RateLimitInputDStream` or any transformed DStream.

File: streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala (hunk 20 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala]: That is because you were running manual clock through the conf generated by TestSuiteBase. When on manual clock, unless you manually add time to the clock, batches will not be generated. If you run on a normal clock, and use eventually sufficiently large timeout, batches will complete.

**Comment by tdas** [comment]
@dragos I didnt delete any comment! The comments may have been to earlier threads which are currently hidden due to outdated diffs. 

I know that these TestSuiteBase are the helper methods that I have provided with Spark Streaming. They were done a long time ago and since then we have learnt quite a bit about the shortcoming of those methods (`runStreams`), and when to use them and when its cleaner to not use them. The problem with `runStreams` is that 
1. it hides part of the management of life cycle of the context. It starts and stop the StreamingContext, which give little control when you want to optimize stuff like.
 - Share SparkContexts across StreamingContexts is not possible through this as the Improves test times if we dont have to spin up SparkContext for every unit test.
 - Tests have unnecessary delays when designed using `runStreams`. This you see for yourself. With or without `runStreams` you have to use `eventually` to test your condition. So why wait for the runStreams to complete (especially when running a real clock) and then test with `eventually`? And how do you decide how many batches to run in `runStreams`? Is it even important to care about the number of batches to run considering the test really wants to verify some other condition being eventually true? Also why go to the complexity of attaching a `TestOutputDStream` if the test does care about the output? It will improve test times as well simplify the test if in these cases you just test with `eventually` alone, without using abstractions that does not really help.

Look at all the newly updated streaming testsuitesl, none of them use `runStreams`. For a reason. Hope this clarifies why I am trying to move the code away from `runStreams`.

**Comment by dragos** [comment]
> On 28 iul. 2015, at 21:42, Tathagata Das <notifications@github.com> wrote:
> 
> @dragos I didnt delete any comment! The comments may have been to earlier threads which are currently hidden due to outdated diffs.
> 
Weird. Probably you attached it to a commit that went away. They seemed to refer to code that was no longer there, but still new. 
> I know that these TestSuiteBase are the helper methods that I have provided with Spark Streaming. They were done a long time ago and since then we have learnt quite a bit about the shortcoming of those methods (runStreams), and when to use them and when its cleaner to not use them. 
> 
How about you deprecate them? It would save time for everyone touching spark streaming. 
> Tests have unnecessary delays when designed using runStreams. This you see for yourself. With or without runStreams you have to use eventually to test your condition. So why wait for the runStreams to complete (especially when running a real clock) and then test with eventually?
If you run on a real clock, I don't see what you gain in terms of time. It's not faster, is it?
> And how do you decide how many batches to run in runStreams?
It's quite simple: in my test I have 3 rates I want to observe, so I need at least 3 batches.
> Is it even important to care about the number of batches to run considering the test really wants to verify some other condition being eventually true?
Yes, since I need to make sure I observe the right number of rate updates, with the right values. 
> Also why go to the complexity of attaching a TestOutputDStream if the test does care about the output?
The 'complexity' we're talking about is 2 lines of code. I'm happy to be proven wrong, but inlining the useful parts of runStreams will be more than that. 

> It will improve test times as well simplify the test if in these cases you just test with eventually alone, without using abstractions that does not really help.
I think they help, but again, happy to be proven wrong. 

How about you rewrite the last test and compare? If you're still convinced it's better I'll rewrite the rest.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+      @volatile var done = false
+      runInBackground {
+        while (!done) {
+          try {
+            dstream.getCurrentRateLimit.foreach(observedRates += _)
+          } catch {
+            case NonFatal(_) => () // don't stop if the executor wasn't installed yet
+          }
+          Thread.sleep(20)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: This is going to be a flaky test. There is no guarantee that this 20 ms sleep will be 20 ms. Can very well be 100 ms. And it cannot be guaranteed that will always get a the 100, 200, 300 will be received by receiver by the time 4 batches are over through `runStreams`. This going to be SUPER FLAKY. The non-flaky way of doing this 
1. run streamingContext normally in real time (no real need for `runStreams`)
2. `eventually { observedRates should contain theSameElementsAs (rates :+ Long.MaxValue) }`
3. `observedRates` must be a synchronized hashset.

**Comment by tdas** [comment]
1. Cannot deprecate TestSuiteBase as it is still perfectly suited for testsuites like `BasicOperationsSuite` where the exact per-batch output needs to be verified. Its suitable for some use cases, just not all usecase, especially not for those where the output is not important.

2. I visited the new tests in ReceiverTrackerSuite again
(a)"rate controller publishes updates": No need for runStreams on exact number of batches, all you need is `eventually { publishCalls > 0 }`
(b) "publish rates reach receivers": No need for runStreams on exact number of batches, all you need is `eventually {dstream.getCurrentRateLimit === Some(200) }`
(c) You are trying to run 4 batches and expect there will 3 observed rates immediately as the 4 batches complete. This CANNOT be guaranteed. Thread.sleep cannot guarantee that you observe the rate limits at the right times so that you observe all the different rates. See comments there, I have suggest how to make this non-flaky. Again, does not require runStreams with exact number of batches.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 23]
```diff
@@ -81,17 +89,25 @@ private class RateLimitInputDStream(@transient ssc_ : StreamingContext)
...
+  def reset(): Unit = {
+    executor_ = null
+  }
+}
 
 /**
  * Dummy receiver implementation
  */
-private class DummyReceiver(receiverId: Int, host: Option[String] = None)
+private[streaming] class DummyReceiver(receiverId: Int, host: Option[String] = None)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Since this receiver and associated InputDStream is used in multiple testsuites, mind renaming this to a more contextually relevant name than `Dummy`? How about `RateTestReceiver` and `RateTestInputDStream`?

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: Also you can augment these classes to count the number of times rates have been received, so that the `MockRateLimitDStream` can be removed. Trying to minimize the number of new classes we create.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala (hunk 22 above)
Comment (dragos) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala]: Ok, in the interest of closing this PR I will do the changes you request ASAP.

However, in the interest of understanding,  I ran it 100 times without it failing once. If you have evidence of it being **SUPER FLAKY** in practice, do you mind sharing it here?

I guess the reason I'm not seeing these failures is that even if theoretically the delay could be higher, say 100ms, since the batch interval is 1s, you'd still sample it 10 times. And looking at the implementation of `SystemClock`, it's still based on `Thread.sleep`, so any systematic delay (say, system under heavy load) will affect both this test and the batch interval sampling rate.

BTW, `eventually` is using `Thread.sleep` as well (15ms between attempts, by default), so I guess the same considerations regarding flakiness should apply.

**Comment by dragos** [comment]
A summary of my changes and a few notes, in the hope of speeding up the process:

- tests are bit faster after moving away from `runStreams`, so I stand corrected
- Tests are still extending `TestSuiteBase` because `withStreamingContext` is defined in there
- I moved the sampling of `observedRates` inside the `eventually` loop, as I don't know how to do it otherwise without using another thread
- I didn't make the `HashSet` synchronized, since there's no background thread anymore

The rest of the changes should match your requests and I don't list them here.

**Comment by tdas** [comment]
LGTM There are still an issues that occurred to me in the last moment, but I am inclined to merge this as is to unblock other PRs - RateController should not be serialized with checkpoints. Does not make sense to recover the RateController. Can you make a PR for this?

Regarding flakiness, it often times occurs only in Jenkins as things can run very slowly in Jenkins, and causes all this Thread.sleep related flakiness to occur. But your current method without background thread make it less flaky, but still does not guarantee. I thought about the test. The least flaky way to test that multiple rate update reach the receiver, would be to send one rate update, wait for it to be received, then send another update with different rate, then wait for it.

**Comment by tdas** [comment]
Aright I merged this. Thanks all!

**Comment by dragos** [comment]
>LGTM There are still an issues that occurred to me in the last moment, but I am inclined to merge this as is to unblock other PRs - RateController should not be serialized with checkpoints. Does not make sense to recover the RateController. Can you make a PR for this?

I forgot to mention this. I spent some time on it, but it looked like the gains were not worth it. The problem is deserialization, and the specific `RateEstimator` instance. At deserialization time there's no `ssc` available yet (it's `null`), so there's no easy way to get to a `SparkConf` in order to get the configured estimator. We could work around that by attaching the controller at a later point, during `start` or some other lifecycle event, but it seemed to complicate matters quite a bit, for little gain that I can see. Thanks for not blocking the PR over this!
